syntax = "proto3";

package ai.lzy.v1.longrunning;

option java_outer_classname = "LongRunning";
option java_package = "ai.lzy.v1.longrunning";

import "google/protobuf/timestamp.proto";
import "google/protobuf/any.proto";
import "google/rpc/status.proto";

service LongRunningService {
  rpc Get (GetOperationRequest) returns (Operation) {}

  rpc Cancel (CancelOperationRequest) returns (Operation) {}

  rpc List (ListOperationsRequest) returns (ListOperationsResponse) {}
}

message Operation {
  string id = 1;
  string description = 2;
  google.protobuf.Timestamp created_at = 3;
  string created_by = 4;
  google.protobuf.Timestamp modified_at = 5;

  bool done = 6;

  google.protobuf.Any metadata = 7;

  oneof result {
    google.rpc.Status error = 8;
    google.protobuf.Any response = 9;
  }
}

message GetOperationRequest {
  string operation_id = 1;
}

message CancelOperationRequest {
  string operation_id = 1;
  string message = 2;
}

message ListOperationsRequest {
  string resource_id = 1;
  int64 page_size = 2;
  string page_token = 3;
  string filter = 4;
}

message ListOperationsResponse {
  repeated Operation operations = 1;
  string next_page_token = 2;
}
